apps/lib/cmp_mock_srv.c

```diff
@@ -183,6 +183,38 @@ int ossl_cmp_mock_srv_set_checkAfterTime(OSSL_CMP_SRV_CTX *srv_ctx, int sec)
     return 1;
 }
 
+/* determine whether to delay response to (non-polling) request */
+static int delayed_delivery(OSSL_CMP_SRV_CTX *srv_ctx, const OSSL_CMP_MSG *req)
+{
+    mock_srv_ctx *ctx = OSSL_CMP_SRV_CTX_get0_custom_ctx(srv_ctx);
+    int req_type = OSSL_CMP_MSG_get_bodytype(req);
+
+    if (ctx == NULL || req == NULL) {
+        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT);
+        return -1;
+    }
+
+    /*
+     * For ir/cr/p10cr/kur delayed delivery is handled separately in
+     * process_cert_request
+     */
+    if (req_type == OSSL_CMP_IR
+        || req_type == OSSL_CMP_CR
+        || req_type == OSSL_CMP_P10CR
+        || req_type == OSSL_CMP_KUR
+        /* Client may use error to abort the ongoing polling */
+        || req_type == OSSL_CMP_ERROR)
+        return 0;
+
+    if (ctx->pollCount > 0 && ctx->curr_pollCount == 0) {
+        /* start polling */
+        if ((ctx->req = OSSL_CMP_MSG_dup(req)) == NULL)
+            return -1;
```

Comment:
you are asserting on this in the caller. Maybe not a good idea for this case?

Reply:
Can you please expand what you mean?

Reply:
You are using ossl_assert(ret != -1) ;
This seems fine for bad input. I am not sure it is fine for a malloc failure.. A memory failure test would die in debug mode with it written like this.

Reply:
Indeed the reaction on malloc failure should not depend on debug mode.

Thinking about the interaction with the caller, which is in `delayed_delivery()` of `cmp_server.c`,
I realized that it should better not use an assertion to handle error cases, 
also because the lib code cannot know under which conditions the called function (of type `OSSL_CMP_SRV_delayed_delivery_cb_t`) in the application throws an error.
I changed it to report any errors thrown by the app.